### PR TITLE
Rule Factory

### PR DESCRIPTION
**Pull request by chrissm79**
This PR adds a new `@rules` directive that can be applied to both mutation arguments as well as nested input object fields.

The directive takes advantage of schema manipulation to flatten out nested rules onto the mutation field which will be run before processing a mutation and provide Laravel validation error(s) in the response meta.

If the `@rules` directive is placed on list, the path will automatically include the `*` notation used by Laravel's validator.

Example Usage:

```graphql
input AddressInput {
  street: String @rules(apply: ["required"])
}

input UserInput {
  name: String @rules(apply: ["required"])
  email: String @rules(apply: ["required", "email"])
  address: [AddressInput]
}

type Mutation {
  createUser(input: UserInput): String
}
```

Will automatically manipulate the schema and place the following directives on the mutation field:

```graphql
type Mutation {
  createUser(input: UserInput): String
    @rules(path: "input.name", rules: ["required"])
    @rules(path: "input.email", rules: ["required", "email"])
    @rules(path: "input.address.*.street", rules: ["required"])
}
```

### CODE REVIEW

**Comment by spawnia** [comment]
This approach works fine for relatively simple Inputs. It does fall apart when dealing with Input Objects that are either really large and heavily nested, or contain circular references.

Because of this, i think this can be a temporary solution, but will have to be changed at a later point. I think the SDL part is great, so we could safely merge this for now and improve on it later.

**Comment by spawnia** [comment]
Oh actually, the generated mutation is not valid. The SDL prohibits the same directive being applied multiple times to the same location.

This might execute fine but will be a problem when we try to validate the schema.

**Comment by chrissm79** [comment]
@spawnia Still working on implementation but can you provide an example of a really large or heavily nested input object(s) that would fail? That way I can write a test for it. As for the circular reference I believe I have a resolution for that. Thanks!

**Comment by spawnia** [comment]
@chrissm79 I will not spell out an example here, as it is pretty tedious to do. Basically, we have mutations in our schema that have ~10 possible child mutations. Each of those can have about the same number of nested mutations itself.

This leads to the rules array growing exponentially in size. This by itself does not guarantee failure, but might cause PHP to run out of memory. I managed to use up > 1GB at the 7th layer of nested mutations.

File: src/Schema/AST/ASTBuilder.php [hunk 1]
```diff
@@ -120,6 +123,45 @@ protected static function applyFieldManipulators(DocumentAST $document)
         }, $document);
     }
 
+    /**
```
Comment (spawnia) [inline src/Schema/AST/ASTBuilder.php]: Clean stuff, really like this!

File: src/Schema/AST/ASTHelper.php [hunk 2]
```diff
@@ -29,4 +31,16 @@ public static function mergeNodeList($original, $addition)
...
     }
+
+    /**
+     * Clone definition node.
+     *
+     * @param Node $node
+     *
+     * @return Node
+     */
+    public static function cloneDefinition(Node $node)
```
Comment (spawnia) [inline src/Schema/AST/ASTHelper.php]: That is in #160 right?

File: src/Schema/AST/DocumentAST.php [hunk 3]
```diff
@@ -134,6 +134,16 @@ public function objectTypes()
         return $this->definitionsByType(ObjectTypeDefinitionNode::class);
     }
 
+    /**
+     * Get all definitions for input types.
+     *
+     * @return Collection
+     */
+    public function inputTypes()
```
Comment (spawnia) [inline src/Schema/AST/DocumentAST.php]: Maybe we should clean up the naming with these functions. We always have `FooBarDefinitionNode` as the underlying type.

File: src/Schema/DirectiveRegistry.php [hunk 4]
```diff
@@ -356,15 +373,15 @@ public function argMiddleware(InputValueDefinitionNode $arg)
...
     /**
-     * @param Directive           $directive
-     * @param FieldDefinitionNode $fieldDefinition
+     * @param Directive $directive
+     * @param Node      $node
      *
      * @return Directive
      */
-    protected function hydrate(Directive $directive, FieldDefinitionNode $fieldDefinition)
+    protected function hydrate(Directive $directive, Node $node)
```
Comment (spawnia) [inline src/Schema/DirectiveRegistry.php]: Hydration can be applied to all directives as of #154 so maybe rebase on that?

File: src/Schema/Directives/Args/RulesDirective.php [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+use Nuwave\Lighthouse\Schema\Factories\RuleFactory;
+use Nuwave\Lighthouse\Schema\Values\ArgumentValue;
+use Nuwave\Lighthouse\Support\Contracts\ArgMiddleware;
+use Nuwave\Lighthouse\Support\Contracts\InputValueManipulator;
+use Nuwave\Lighthouse\Support\Traits\HandlesDirectives;
+
+class RulesDirective implements ArgMiddleware, InputValueManipulator
+{
+    // TODO: Remove this need for this
+    use HandlesDirectives;
```
Comment (spawnia) [inline src/Schema/Directives/Args/RulesDirective.php]: No longer needed once we have #154

File: src/Schema/Factories/RuleFactory.php [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+                $field->arguments = new NodeList($arguments);
+
+                return $field;
+            })->toArray();
+
+        $mutation->fields = new NodeList($fields);
+
+        // NOTE: This is currently not needed because we've already
+        // mutated the fields on the original instance.
+        // $documentAST->setDefinition($mutation);
```
Comment (spawnia) [inline src/Schema/Factories/RuleFactory.php]: Still true, now that you use `cloneDefinition()` ?

**Pull request closed without merging**